### PR TITLE
REST for OneToOne, OneToMany, ManyToMany.

### PR DESCRIPTION
**Pull request by dkorolev**

### CODE REVIEW

File: Bricks/util/iterator.h [hunk 1]
```diff
@@ -32,12 +32,12 @@ namespace current {
 
 template <typename MAP>
 struct GenericMapIterator final {
-  using iterator_t = typename MAP::const_iterator;
+  using inner_iterator_t = typename MAP::const_iterator;
```
Comment (mzhurovich) [inline Bricks/util/iterator.h]: s/inner_iterator_t/map_const_iterator_t/?

File: Bricks/util/iterator.h [hunk 2]
```diff
@@ -49,7 +49,8 @@ struct GenericMapIterator final {
 
 template <typename MAP>
 struct GenericMapAccessor final {
-  using iterator_t = GenericMapIterator<MAP>;
+  using const_iterator = GenericMapIterator<MAP>;
+  using iterator_t = const_iterator;
```
Comment (mzhurovich) [inline Bricks/util/iterator.h]: This is pretty unobvious :-/

File: Storage/semantics.h [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+namespace rest {
+struct RESTWithSingleKey {};
+struct RESTWithPairedKey {};
+}  // namespace current::storage::semantics::rest
+
+namespace primary_key {
+struct Key {};
+struct RowCol {};
+}  // namespace current::storage::semantics::primary_key
```
Comment (mzhurovich) [inline Storage/semantics.h]: Don't the structures described in `rest` and in `primary_key` actually mean the same?

File: Storage/api.h [hunk 4]
```diff
@@ -124,31 +127,29 @@ struct PerFieldRESTfulHandlerGenerator {
...
-    using PUTHandler = DataHandlerImpl<PUT, specific_field_t, entry_t, key_t>;
-    using DELETEHandler = DataHandlerImpl<DELETE, specific_field_t, entry_t, key_t>;
+    using top_level_operation_t = semantics::rest::operation::top_level_operation_for_field_t<specific_field_t>;
+
+    using GETHandler = DataHandlerImpl<GET, top_level_operation_t, specific_field_t, entry_t, key_t>;
+    using POSTHandler = DataHandlerImpl<POST, top_level_operation_t, specific_field_t, entry_t, key_t>;
+    using PUTHandler = DataHandlerImpl<PUT, top_level_operation_t, specific_field_t, entry_t, key_t>;
+    using DELETEHandler = DataHandlerImpl<DELETE, top_level_operation_t, specific_field_t, entry_t, key_t>;
 
     using SchemaRoutesGenerator = SchemaHandlerImpl<entry_t>;
```
Comment (mzhurovich) [inline Storage/api.h]: Move this to line 287?

File: Storage/test.cc [hunk 5]
```diff
@@ -23,17 +23,23 @@ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.
 *******************************************************************************/
 
+// #define STORAGE_ONLY_RUN_RESTFUL_TESTS
```
Comment (mzhurovich) [inline Storage/test.cc]: Remove?

File: Storage/test.cc [hunk 6]
```diff
@@ -2113,15 +2130,43 @@ TEST(TransactionalStorage, RESTfulAPITest) {
               HTTP(GET(base_url + "/api/data/like?row=max&col=beer")).body);
     EXPECT_EQ("{\"row\":\"max\",\"col\":\"beer\",\"details\":\"Cheers!\"}\n",
               HTTP(GET(base_url + "/api/data/like?key1=max&key2=beer")).body);
-    // Meh, this test is not for `AdvancedHypermedia`. -- D.K.
+    // Meh, this test is not for the real Hypermedia. -- D.K.
+    // TODO(dkorolev: Make is so?
```
Comment (mzhurovich) [inline Storage/test.cc]: TODO(dkorolev**)**: Make i**t** so?

Comment (dkorolev) [inline Storage/test.cc]: :+1: 21c07c2
And added a `DIMA_FIXME` to actually test this.

File: Bricks/util/iterator.h (hunk 1 above)
Comment (dkorolev) [inline Bricks/util/iterator.h]: True indeed. Let's leave renaming to the last pass. Added a `DIMA_FIXME` for now to grep for it later, won't merge until we clean those up.

File: Bricks/util/iterator.h (hunk 2 above)
Comment (dkorolev) [inline Bricks/util/iterator.h]: Same as above -- suggest we leave renaming to the last pass, added a DIMA_FIXME to grep for it later.

File: Storage/semantics.h (hunk 3 above)
Comment (dkorolev) [inline Storage/semantics.h]: One is container-level (internal representation), one is exposure-level (HTPT/REST). I'll get these to `enum class`-es soon.

File: Storage/api.h (hunk 4 above)
Comment (dkorolev) [inline Storage/api.h]: :+1: 21c07c2

Interestingly, I've moved it down, and then up again, in hopes to minimize the diff size of this CR. Fixed, and, while on this note, indented the large lambda above as well.

File: Storage/test.cc (hunk 5 above)
Comment (dkorolev) [inline Storage/test.cc]: Added `DIMA_FIXME`, will remove before merging.

**Comment by dkorolev** [comment]
@mzhurovich -- cleaned up simple things. Remaining for this PR:
1. Implement and test the `"brief"` mode.
2. Test DELETE for REST matrix.
3. Take care of `DIMA_FIXME`-s.
PTAL && thanks!

**Comment by dkorolev** [comment]
@mzhurovich Looks Green To Me, PTAL!

File: Storage/rest/hypermedia.h [hunk 7]
```diff
@@ -40,28 +40,42 @@ namespace current {
...
-      const std::string& url, const std::string& url_directory, const T& data, bool set_success = false)
-      : url(url), url_full(url + "?full"), url_brief(url + "?brief"), url_directory(url_directory), data(data) {
-    if (set_success) {
-      success = true;
-    }
+#ifdef MACRO_STRUCT_BODY
+#error "`MACRO_STRUCT_BODY` should not be defined."
+#endif
+
+#define MACRO_STRUCT_BODY(struct_name) \
```
Comment (mzhurovich) [inline Storage/rest/hypermedia.h]: s/MACRO_STRUCT_BODY/REST_RESPONSE_STRUCT_BODY/?

File: Storage/test.cc [hunk 8]
```diff
@@ -2159,6 +2236,240 @@ TEST(TransactionalStorage, RESTfulAPITest) {
...
+
+  using namespace transactional_storage_test;
+  using namespace current::storage::rest;
+  using Storage = SimpleStorage<SherlockInMemoryStreamPersister>;
+
+  Storage storage;
+
+  const auto base_url = current::strings::Printf("http://localhost:%d", FLAGS_transactional_storage_test_port);
+
+  const auto rest1 = RESTfulStorage<Storage>(storage, FLAGS_transactional_storage_test_port, "/text", "");
```
Comment (mzhurovich) [inline Storage/test.cc]: s/text/plain/?

File: Storage/test.cc [hunk 9]
```diff
@@ -2159,6 +2236,240 @@ TEST(TransactionalStorage, RESTfulAPITest) {
...
+  const auto rest2 = RESTfulStorage<Storage, current::storage::rest::Simple>(
```
Comment (mzhurovich) [inline Storage/test.cc]: s/basic/simple/?

File: Storage/test.cc [hunk 10]
```diff
@@ -2159,6 +2236,240 @@ TEST(TransactionalStorage, RESTfulAPITest) {
...
+    EXPECT_EQ(201,
+              static_cast<int>(HTTP(PUT(base_url + "/basic/data/composite_m2m/!3/1",
+                                        SimpleComposite("!3", std::chrono::microseconds(1)))).code));
+    EXPECT_EQ(201,
+              static_cast<int>(HTTP(PUT(base_url + "/hypermedia/data/composite_m2m/!3/2",
+                                        SimpleComposite("!3", std::chrono::microseconds(2)))).code));
+  }
+
+  {
+    // Browse the collection in various ways using the `Text` API.
```
Comment (mzhurovich) [inline Storage/test.cc]: s/Text/Plain/?

File: Storage/test.cc [hunk 11]
```diff
@@ -2159,6 +2236,240 @@ TEST(TransactionalStorage, RESTfulAPITest) {
...
+    }
+    {
+      const auto response = HTTP(GET(base_url + "/text/data/composite_m2m.col/3"));
+      EXPECT_EQ(200, static_cast<int>(response.code));
+      EXPECT_EQ("{\"row\":\"!1\",\"col\":3}\n{\"row\":\"!2\",\"col\":3}\n", response.body);
+    }
+  }
+
+  {
+    // Browse the collection in various ways using the `Hypermedia` API.
```
Comment (mzhurovich) [inline Storage/test.cc]: s/Hypermedia/Simple/

File: Storage/api.h [hunk 12]
```diff
@@ -124,163 +127,237 @@ struct PerFieldRESTfulHandlerGenerator {
...
+                                 // Capture local variables by value for safe async transactions.
+                                 [&storage, handler, generic_input, &field, url_key, field_name](
+                                     immutable_fields_t fields) -> Response {
+                                   using RowGETInput =
+                                       RESTfulGETRowColInput<STORAGE,
+                                                             typename PARTIAL_KEY_OPERATION::key_completeness_t,
+                                                             specific_field_t>;
+                                   RowGETInput input(
+                                       std::move(generic_input), fields, field, field_name, url_key);
+                                   return handler.Run(input);
```
Comment (mzhurovich) [inline Storage/api.h]: Just a thought - do we want, theoretically, to receive this input by rvalue?

File: Storage/test.cc (hunk 8 above)
Comment (dkorolev) [inline Storage/test.cc]: :+1: b83354be616d8b6a386e0c3eb9cb32e7eb30f00c

File: Storage/test.cc (hunk 9 above)
Comment (dkorolev) [inline Storage/test.cc]: :+1: b83354be616d8b6a386e0c3eb9cb32e7eb30f00c

File: Storage/test.cc (hunk 10 above)
Comment (dkorolev) [inline Storage/test.cc]: :+1: b83354be616d8b6a386e0c3eb9cb32e7eb30f00c

File: Storage/test.cc (hunk 11 above)
Comment (dkorolev) [inline Storage/test.cc]: :+1: b83354be616d8b6a386e0c3eb9cb32e7eb30f00c

File: Storage/api.h (hunk 12 above)
Comment (dkorolev) [inline Storage/api.h]: No :) changed `*Input input`-s into `const`. There's `context` for user data.

File: Current.vcxproj.filters [hunk 13]
```diff
@@ -1066,22 +1066,31 @@
...
+      <Filter>Third Party</Filter>
+    </ClInclude>
+    <ClInclude Include="Storage\rest\simple.h">
       <Filter>Third Party</Filter>
     </ClInclude>
     <ClInclude Include="Storage\rest\hypermedia.h">
       <Filter>Third Party</Filter>
     </ClInclude>
-    <ClInclude Include="Storage\rest\sfinae.h">
+    <ClInclude Include="Storage\base.h">
```
Comment (mzhurovich) [inline Current.vcxproj.filters]: This one was included in the line 898

**Comment by mzhurovich** [comment]
Помолясь применимъ!

**Pull request merged**